### PR TITLE
Perform reciprocal optimization with foreach_div

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #128433

Fixes https://github.com/pytorch/pytorch/issues/114165

Internal xref
https://fb.workplace.com/groups/1144215345733672/posts/2801223606699496/

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

### CODE REVIEW

File: aten/src/ATen/native/cuda/ForeachBinaryOpScalar.cu [hunk 1]
```diff
@@ -191,11 +191,44 @@ std::vector<Tensor> foreach_scalar_pow_list_kernel_cuda(
...
 // Currently multi tensor apply can only return result of the same type as
 // input.
-FOREACH_BINARY_OP_SCALAR(
-    all_types_complex_bool_half_bfloat16,
-    div,
-    std::divides,
-    /*div_op*/ true);
+//
+// Implement via multiply with reciprocal as it's faster.  Loses one bit of
+// precision.
```
Comment (awgu) [inline aten/src/ATen/native/cuda/ForeachBinaryOpScalar.cu:196]: nit: Do you think it is worth mentioning that this is the same behavior as `torch.div_`?

**Review by awgu — APPROVED** [review]
This sounds good to me! Feel free to get feedback from the others though.

**Comment by awgu** [comment]
Related: https://github.com/pytorch/pytorch/issues/114165

File: aten/src/ATen/native/cuda/ForeachBinaryOpScalar.cu [hunk 2]
```diff
@@ -191,11 +191,47 @@ std::vector<Tensor> foreach_scalar_pow_list_kernel_cuda(
...
+  all_types_complex_bool_half_bfloat16_<std::multiplies>(
+      tensors, scalar_reciprocal(scalar));
+}
+
+std::vector<Tensor> foreach_tensor_div_scalar_kernel_cuda(
+    TensorList tensors,
+    const Scalar& scalar) {
+  check_foreach_api_restrictions(tensors);
+  if (!can_use_fast_route(tensors, scalar, true)) {
+    return at::native::foreach_tensor_div_scalar_kernel_slow(
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/ForeachBinaryOpScalar.cu:228]: OOPS

**Pull request closed without merging**